myapp/README.md

```diff
@@ -0,0 +1,242 @@
+# Restful API for Wiredcraft User Table
+
+## Running Environment:
+
+* Nodejs 14.16.0
+* Redis 6.2.0
+
+## Install Steps:
+
+* ```git clone https://github.com/eronekogin/test-backend.git```
+* ```cd myapp```
+* ```npm install```
+* Customizing the default.json file under ./config folder, an example is as follows:
+```json
+{
+    "User": {
+        // Set up redis connections, currently user name and password are not supported. 
+        "dbConfig": {
+            "host": "127.0.0.1",
+            "port": "6379"
+        },
+
+        // Set up table names, you could customize it to any string you like.
+        "USER_HASH_PREFIX": "userId",
+        "USER_ID_SORTED_SET": "userIds",
+        "USER_HASH_PREFIX": "userId",
+        "FOLLOWING_HASH_PREFIX": "followings",
+        "FOLLOWER_HASH_PREFIX": "followers",
+        "FRIENDS_HASH_PREFIX": "friends",
+        "NEAR_FRIENDS_HASH_PREFIX": "nearFriends",
+        "NEAR_USERS_HASH_PREFIX": "nearUsers",
+        "ADDRESS_HASH_PREFIX": "address",
+        "USER_ID_SORTED_SET": "userIds",
+        "GEO_SORTED_SET": "geos",
+
+        // Currently modifying allowed input fields are not supported.
+        "USER_ALLOWED_INPUT_FIELDS": [
+            "name",
+            "dob",
+            "address",
+            "description"
+        ]
+    }
+}
+```
+* Make sure you could connect to the above configured redis server.
+* ```npm start```
+* Then you will see some similar messages as follows:
+```test
+> myapp@1.0.0 start D:\nodejs\workspace\test-backend\myapp
+> node ./bin/www
+
+D:\nodejs\workspace\test-backend\myapp\bin
+```
```

Comment:
with `User.dbConfig.host: '192.168.0.107'`, the app cannot handle incoming request, 
are there other configurations that need to update to show the error details?

Reply:
* As specified in the myapp/README.md file, the person who tries to deploy this application should make sure they are able to connect to the redis server specified by the host and port name.
* Currently the application does not provide any error message when we cannot connect to the remote redis server when starting up the application. If the database configuration is incorrect, it will hang.
* An enhancement for fixing this issue is to add a try/catch check when trying to create the redis client in the myapp/databases/redisClient.js module so that when the client creation is failing, an error will be sent back to the user console.

Reply:
will the redis client retry to connect when there are network interruptions, 
what can we do to help detect failures and recover quickly in production environment?

Reply:
* The current version of this application is using node-redis^3.0.2, where the redis.createClient() method will create a client instance that will auto reconnecting to the specified redis server by default.
* For production environment, it is not good enough to just use the default parameters set by the redis.createClient() method but instead we should offer the retry_strategy with the specific requirement from the production environment, an example is as follows:

```javascript
const client = redis.createClient({
  retry_strategy: function(options) {
    if (options.error && options.error.code === "ECONNREFUSED") {
      // End reconnecting on a specific error and flush all commands with
      // a individual error
      return new Error("The server refused the connection");
    }
    if (options.total_retry_time > 1000 * 60) {
      // End reconnecting after a specific timeout and flush all commands
      // with a individual error, currently the limitation is 1 minute.
      return new Error("Retry time exhausted");
    }
    if (options.attempt > 10) {
      // End reconnecting with built in error, currently the limit is 10 times.
      return undefined;
    }
    // reconnect after a specific time.
    return Math.min(options.attempt * 100, 3000);
  },
});
```
  * The above strategy will prevent the application from holding up due to losing connection to the redis-server, and the customized error types could be caught by the application so that it could log the target error and send the response to the user without being frozen forever.
* Meanwhile, we could also encapsulate the client object as a proxy, which functions as follows:
  * When the actual redis client is not ready due to network broken or remote server error, the proxy object will have the same interfaces as a RedisClient object, but does no real work.
  * Once the actual redis client is ready, either initial connecting or recovered from disconnection, we set the proxy object to it and now the proxy object could do some real database work like sending commands to redis.

Reply:
Application that can automatically recover from short period interruptions will save a lot of DevOps effort.
When auto recovery becomes more and more complex, one disadvantage is that some edge cases are hard to test, we won't have confidence about what will happen in edge cases.
Another way to help recover is to use monitor, app can send health status to monitor (or let monitor pull health status from app), when app is unhealthy for some time, DevOps can receive alert and manually investigate & solve the issue.